File: packages/@livestore/wa-sqlite/src/libsession.js
// This file should be included in the build with --post-js.

;(function () {
  const AsyncFunction = Object.getPrototypeOf(async function () {}).constructor

  /**
   * Apply a changeset to the "main" database, relaying filter and conflict
   * callbacks to JavaScript through the C adapter layer via method dispatch.
   *
   * @param {number} db - Database handle pointer.
   * @param {number} nChangeset - Size of changeset in bytes.
   * @param {number} pChangeset - Pointer to changeset blob in WASM memory.
   * @param {function(string): number | null} xFilter - Optional filter callback.
   *   Called with the table name for each table in the changeset. Return 0 to
   *   skip changes for that table, non-zero to include them. Pass null to
   *   include all tables.
   * @param {function(number): number} xConflict - Conflict handler called with
   *   the conflict type (SQLITE_CHANGESET_DATA, _NOTFOUND, _CONFLICT,
   *   _CONSTRAINT, or _FOREIGN_KEY). Must return SQLITE_CHANGESET_OMIT (0),
   *   SQLITE_CHANGESET_REPLACE (1), or SQLITE_CHANGESET_ABORT (2).
   *   See https://sqlite.org/session/sqlite3changeset_apply.html
   * @returns {number} SQLITE_OK on success, or an SQLite error code.
   *
   * @see https://sqlite.org/session/sqlite3changeset_apply.html
   */
  Module['changeset_apply'] = function (db, nChangeset, pChangeset, xFilter, xConflict) {
    const pAsyncFlags = Module['_sqlite3_malloc'](4)
    let asyncFlags = 0
    if (xFilter && xFilter instanceof AsyncFunction) asyncFlags |= (1 << 0)
    if (xConflict instanceof AsyncFunction) asyncFlags |= (1 << 1)
    setValue(pAsyncFlags, asyncFlags, 'i32')

    const target = {}
    if (xFilter) {
      target.xFilter = (zTab) => xFilter(Module['UTF8ToString'](zTab))
    }
    target.xConflict = (eConflict, _pIter) => xConflict(eConflict)

    Module['setCallback'](pAsyncFlags, target)

    const result = ccall(
      'libsession_changeset_apply', 'number',
      ['number', 'number', 'number', 'number', 'number', 'number'],
      [db, nChangeset, pChangeset, xFilter ? 1 : 0, 1, pAsyncFlags]
    )

    Module['deleteCallback'](pAsyncFlags)
    Module['_sqlite3_free'](pAsyncFlags)

    return result
  }
})()

File: packages/@livestore/wa-sqlite/src/libsession.c
#include <stdio.h>
#include <emscripten.h>
#include <sqlite3.h>

#include "libadapters.h"

enum {
  xFilter,
  xConflict,
};

#define SESSION_JS(SIGNATURE, KEY, METHOD, ...) \
  (asyncFlags & (1 << METHOD) ? \
    SIGNATURE##_async(KEY, #METHOD, __VA_ARGS__) : \
    SIGNATURE(KEY, #METHOD, __VA_ARGS__))

/*
** Filter callback that relays to a JavaScript function via method
** dispatch. Returns non-zero to include changes for the table,
** or zero to exclude them.
**
** See https://sqlite.org/session/sqlite3changeset_apply.html
*/
static int libsession_xFilter(void* pCtx, const char* zTab) {
  const int asyncFlags = pCtx ? *(int *)pCtx : 0;
  return SESSION_JS(ippp, pCtx, xFilter, zTab);
}

/*
** Conflict handler callback that relays to a JavaScript function
** via method dispatch.
**
** See https://sqlite.org/session/sqlite3changeset_apply.html
*/
static int libsession_xConflict(void* pCtx, int eConflict, sqlite3_changeset_iter* p) {
  const int asyncFlags = pCtx ? *(int *)pCtx : 0;
  return SESSION_JS(ippip, pCtx, xConflict, eConflict, p);
}

/*
** Apply the changeset (pChangeset, nChangeset) to the "main" database
** of handle db. Both xFilter and xConflict are resolved through the
** adapter relay via method dispatch on the target object stored at
** key pCtx.
**
** Return SQLITE_OK on success, or an SQLite error code on failure.
*/
int EMSCRIPTEN_KEEPALIVE libsession_changeset_apply(
  sqlite3* db,                  /* Database handle */
  int nChangeset,               /* Size of changeset in bytes */
  void* pChangeset,             /* Pointer to changeset blob */
  int hasXFilter,               /* Non-zero if a filter callback is registered */
  int hasXConflict,             /* Non-zero if a conflict handler is registered */
  void* pCtx) {                 /* Callback lookup key */
  return sqlite3changeset_apply(
    db, nChangeset, pChangeset,
    hasXFilter ? &libsession_xFilter : NULL,
    hasXConflict ? &libsession_xConflict : NULL,
    pCtx);
}
